tests/test_utils.py

```diff
@@ -161,19 +253,113 @@ def test_options_for_gather_missing_arg_parameter(monkeypatch, args):
+    command = "./scan --scan=a11y"
+    monkeypatch.setattr(sys, "argv", command.split(" "))
+    with pytest.raises(SystemExit) as exc:
+        subutils.options_for_scan()
+    assert exc.typename == "SystemExit"
+    out, err = capsys.readouterr()
+    assert err.endswith("arguments are required: domains\n")
+
+
+def test_options_for_scan_basic(monkeypatch):
+    command = "./scan example.org --scan=a11y"
+    monkeypatch.setattr(sys, "argv", command.split(" "))
+    result = subutils.options_for_scan()
+    assert result == {
+        "_": default_underscore_scan,
+        "domains": "example.org",
+        "scan": "a11y",
+        "output": "./",
+        **scan_default_false_values,
+    }
+
+
+@pytest.mark.parametrize("args", [
+    "./scan --help",
+    "./scan -h",
+])
+def test_options_for_scan_help(monkeypatch, capsys, args):
+    monkeypatch.setattr(sys, "argv", args.split(" "))
+    # Handling exception here instead of with decorator because we want to
+    # examine the console output.
+    with pytest.raises(SystemExit) as exc:
+        subutils.options_for_scan()
+    assert exc.typename == "SystemExit"
+    out, err = capsys.readouterr()
+    assert out.startswith("usage: scan [-h]")
+
+
+@pytest.mark.parametrize("arg", scan_args_with_mandatory_values)
+@pytest.mark.xfail(raises=argparse.ArgumentError)
+def test_options_for_scan_missing_mandatory(monkeypatch, arg):
+    command = "./scan example.org --scan=a11y --%s" % arg.replace("_", "-")
+    monkeypatch.setattr(sys, "argv", command.split(" "))
+    subutils.options_for_scan()
+    command = "./scan example.org --scan=a11y --%s=" % arg.replace("_", "-")
+    monkeypatch.setattr(sys, "argv", command.split(" "))
+    subutils.options_for_scan()
+
+
+@pytest.mark.xfail(raises=argparse.ArgumentTypeError)
+def test_options_for_scan_lambda_profile_no_lambda(monkeypatch):
+    command = "./scan example.org --scan=a11y --lambda-profile=something"
+    monkeypatch.setattr(sys, "argv", command.split(" "))
+    subutils.options_for_scan()
+
+
+@pytest.mark.parametrize("command,expected", [
+    (
+        "./scan example.org --scan=a11y --workers=1",
+        {
+            "_": default_underscore_scan,
+            "domains": "example.org",
+            "scan": "a11y",
+            "workers": "1",
+            "output": "./",
+            **scan_default_false_values,
+        }
+    ),
+    (
+        "./scan example.org --scan=a11y --output=..",
+        {
+            "_": {
+                "cache_dir": "../cache",
+                "report_dir": "..",
+                "results_dir": "../results",
+            },
+            "domains": "example.org",
+            "scan": "a11y",
+            "output": "..",
+            **scan_default_false_values,
+        }
+    ),
+])
+def test_options_for_scan_check_for_single_args(monkeypatch, command, expected):
+    monkeypatch.setattr(sys, "argv", command.split(" "))
+    result = subutils.options_for_scan()
+    assert result == expected
```

Comment:
I definitely want to get more tests in here, and a few simple tests making sure our command line parsing works properly makes sense. But this seems like a lot of testing code for what in most tools is a routine parsing process that uses some auto-inferrable template for specifying options. I think this is another manifestation of my comment above, that the argument parsing code has gotten much more complex than expected.

Reply:
These tests are in place to cover the functionality that was present before I changed the argument-parsing code; I tried not to add any new functionality, just to make tests that addressed the various edge cases I found in `scan` and the scanners.